### PR TITLE
Properly raise yaml errors when parser fails to read story / rule files / domain

### PR DESCRIPTION
**Pull request by tmbo**
**Issue**:
- validation errors in story / rule files did not lead to the training failing but rather resulted in a userwarning
- syntactically invalid yaml files just got ignored instead of an error getting raised

**Proposed changes**:
- use proper docs link for training data (don't use base url)
- removed unused `update_number_of_epochs` from `tests.utilities`
- `YamlSyntaxException` and `YamlValidationException` inherit from `YamlException` now (to make it easier to catch both of them)
- duplicate keys in a yaml file will raise an exception by default

TODO:
- [x] fail if domain reading fails
- [x] check if tests still pass

### CODE REVIEW

File: rasa/shared/core/domain.py [hunk 1]
```diff
@@ -1197,14 +1198,9 @@ def is_domain_file(filename: Text) -> bool:
 
         if not is_likely_yaml_file(filename):
             return False
-        try:
-            content = rasa.shared.utils.io.read_yaml_file(filename)
-            if any(key in content for key in ALL_DOMAIN_KEYS):
-                return True
-        except YAMLError:
```
Comment (tmbo) [inline rasa/shared/core/domain.py:1204]: this function won't swallow errors anymore

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py [hunk 2]
```diff
@@ -95,15 +99,8 @@ def read_from_string(self, string: Text) -> List[StoryStep]:
         Returns:
             `StoryStep`s read from `string`.
         """
-        try:
-            rasa.shared.utils.validation.validate_yaml_schema(string, CORE_SCHEMA_FILE)
-            yaml_content = rasa.shared.utils.io.read_yaml(string)
-        except (ValueError, ParserError) as e:
```
Comment (tmbo) [inline rasa/shared/core/training_data/story_reader/yaml_story_reader.py:101]: won't swallow errors anymore but instead training will fail

File: rasa/cli/test.py [hunk 3]
```diff
@@ -137,8 +138,8 @@ def run_nlu_test(args: argparse.Namespace) -> None:
                     rasa.shared.utils.io.read_file(file), CONFIG_SCHEMA_FILE,
                 )
                 config_files.append(file)
-            except validation_utils.YamlValidationException:
-                logger.debug(
+            except YamlException:
+                logger.warning(
```
Comment (wochinge) [inline rasa/cli/test.py:142]: should rather be a `UserWarning` no?

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 4]
```diff
@@ -100,7 +100,7 @@ def train(
                 "Can not train an intent classifier as there are not "
                 "enough intents. Need at least 2 different intents. "
                 "Skipping training of intent classifier.",
-                docs=DOCS_URL_TRAINING_DATA_NLU,
+                docs=DOCS_URL_TRAINING_DATA,
```
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:103]: What about still linking to the `NLU Training Data` section?

File: rasa/shared/core/domain.py [hunk 5]
```diff
@@ -22,7 +22,8 @@
 
 import rasa.shared.constants
 import rasa.shared.core.constants
-from rasa.shared.exceptions import RasaException
+from rasa.shared.exceptions import RasaException, YamlException, YamlSyntaxException
```
Comment (wochinge) [inline rasa/shared/core/domain.py:25]: unused imports

File: rasa/shared/core/domain.py [hunk 6]
```diff
@@ -1197,14 +1198,9 @@ def is_domain_file(filename: Text) -> bool:
 
         if not is_likely_yaml_file(filename):
             return False
-        try:
```
Comment (wochinge) [inline rasa/shared/core/domain.py:1200]: this is a boolean check. Isn't it supposed to not throw things? The least we should do is document the exception in the docstring as it's not intuitive.

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py [hunk 7]
```diff
@@ -3,6 +3,7 @@
 from typing import Dict, Text, List, Any, Optional, Union
 
 import rasa.shared.data
+from rasa.shared.exceptions import YamlException
```
Comment (wochinge) [inline rasa/shared/core/training_data/story_reader/yaml_story_reader.py:6]: unused imports

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py [hunk 8]
```diff
@@ -95,15 +99,8 @@ def read_from_string(self, string: Text) -> List[StoryStep]:
         Returns:
             `StoryStep`s read from `string`.
```
Comment (wochinge) [inline rasa/shared/core/training_data/story_reader/yaml_story_reader.py:100]: optional: document the potential exceptions in the docstring

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py [hunk 9]
```diff
@@ -160,16 +157,8 @@ def is_key_in_yaml(cls, file_path: Text, *keys: Text) -> bool:
...
-            return any(key in content for key in keys)
-        except Exception as e:
-            # Using broad `Exception` because yaml library is not exposing all Errors
-            rasa.shared.utils.io.raise_warning(
-                f"Tried to open '{file_path}' and load its data, but failed "
-                f"to read it. There seems to be an error with the yaml syntax: {e}"
-            )
-            return False
+        content = rasa.shared.utils.io.read_yaml_file(file_path)
+        return any(key in content for key in keys)
```
Comment (wochinge) [inline rasa/shared/core/training_data/story_reader/yaml_story_reader.py:161]: same as for the other one: I think these boolean checks should swallow

Comment (wochinge) [reply rasa/shared/core/training_data/story_reader/yaml_story_reader.py:161]: (from the category: things I would have thought I would never say)

File: rasa/shared/exceptions.py [hunk 10]
```diff
@@ -13,18 +13,46 @@ class RasaXTermsError(RasaException):
     """Error in case the user didn't accept the Rasa X terms."""
 
 
-class YamlSyntaxException(RasaException):
+class YamlException(RasaException):
+    """Raised if there is an error reading yaml."""
+
+    def __init__(self, filename: Optional[Text] = None):
```
Comment (wochinge) [inline rasa/shared/exceptions.py:19]:
```suggestion
    def __init__(self, filename: Optional[Text] = None) -> None:
```

File: rasa/shared/exceptions.py [hunk 11]
```diff
@@ -13,18 +13,46 @@ class RasaXTermsError(RasaException):
...
+
+class YamlSyntaxException(YamlException):
     """Raised when a YAML file can not be parsed properly due to a syntax error."""
 
-    def __init__(self, filename: Text, underlying_yaml_exception: Exception):
+    def __init__(
+        self,
+        filename: Optional[Text] = None,
+        underlying_yaml_exception: Optional[Exception] = None,
+    ):
```
Comment (wochinge) [inline rasa/shared/exceptions.py:34]:
```suggestion
    ) -> None:
```

File: rasa/shared/exceptions.py [hunk 12]
```diff
@@ -13,18 +13,46 @@ class RasaXTermsError(RasaException):
...
         self.underlying_yaml_exception = underlying_yaml_exception
-        self.filename = filename
+        super(YamlSyntaxException, self).__init__(filename)
```
Comment (wochinge) [inline rasa/shared/exceptions.py:36]: optional: if possible call the `super()` thing before doing anything else https://www.notion.so/rasa/Python-Code-Conventions-26835ffd72484dd699f28fa693916d2e#498bb338d9ca4de2892a70d0e7cb71d2

File: rasa/shared/exceptions.py [hunk 13]
```diff
@@ -13,18 +13,46 @@ class RasaXTermsError(RasaException):
...
+            self.underlying_yaml_exception.note = None
+            exception_text += f" {self.underlying_yaml_exception}"
+
+        if self.filename:
+            exception_text = exception_text.replace(
+                'in "<unicode string>"', f'in "{self.filename}"'
+            )
+
+        exception_text += (
+            "\n\nYou can use http://www.yamllint.com/ to validate the "
```
Comment (wochinge) [inline rasa/shared/exceptions.py:55]: (it slightly bugs me out that they are not using https)

File: rasa/shared/nlu/training_data/formats/rasa_yaml.py [hunk 14]
```diff
@@ -4,12 +4,12 @@
 from typing import Text, Any, List, Dict, Tuple, Union, Iterator, Optional
 
 import rasa.shared.data
-from rasa.shared.exceptions import RasaException
+from rasa.shared.exceptions import RasaException, YamlException, YamlSyntaxException
```
Comment (wochinge) [inline rasa/shared/nlu/training_data/formats/rasa_yaml.py:7]: unused imports

File: rasa/shared/nlu/training_data/formats/rasa_yaml.py [hunk 15]
```diff
@@ -345,19 +345,9 @@ def is_yaml_nlu_file(filename: Text) -> bool:
...
-        except (YAMLError, Warning) as e:
-            logger.error(
-                f"Tried to check if '{filename}' is a NLU file, but failed to "
-                f"read it. If this file contains NLU data, you should "
-                f"investigate this error, otherwise it is probably best to "
-                f"move the file to a different location. "
-                f"Error: {e}"
-            )
-            return False
+        content = rasa.shared.utils.io.read_yaml_file(filename)
```
Comment (wochinge) [inline rasa/shared/nlu/training_data/formats/rasa_yaml.py:348]: same as the other ones (let's discuss just one of these please 😂 )

File: rasa/shared/utils/io.py [hunk 16]
```diff
@@ -269,7 +271,10 @@ def read_yaml_file(filename: Union[Text, Path]) -> Union[List[Any], Dict[Text, A
     Args:
         filename: The path to the file which should be read.
```
Comment (wochinge) [inline rasa/shared/utils/io.py:272]: optional (as it's not your change): Document the exception in the docstring. (ideally also the `Returns:`)

File: rasa/shared/utils/validation.py [hunk 17]
```diff
@@ -8,13 +9,12 @@
 from ruamel.yaml.constructor import DuplicateKeyError
 
 import rasa.shared
-from rasa.shared.exceptions import RasaException
+from rasa.shared.exceptions import RasaException, YamlException, YamlSyntaxException
```
Comment (wochinge) [inline rasa/shared/utils/validation.py:12]: unused
```suggestion
from rasa.shared.exceptions import YamlException, YamlSyntaxException
```

File: rasa/shared/utils/validation.py [hunk 18]
```diff
@@ -43,17 +43,16 @@ def __init__(
             content: yaml content loaded from the file (used for line information)
         """
         self.message = message
-        self.filename = filename
         self.validation_errors = validation_errors
         self.content = content
-        super(YamlValidationException, self).__init__()
+        super(YamlValidationException, self).__init__(filename)
```
Comment (wochinge) [inline rasa/shared/utils/validation.py:48]: optional: if possible, call the `super()` before the other instructions in the constructor

File: rasa/shared/utils/validation.py [hunk 19]
```diff
@@ -218,18 +207,18 @@ def validate_training_data_format_version(
...
         # not raising here since it's not critical
         logger.warning(
-            f"Training data file {filename} doesn't have a "
-            f"'{KEY_TRAINING_DATA_FORMAT_VERSION}' key. "
+            f"The '{KEY_TRAINING_DATA_FORMAT_VERSION}' key is missing in "
+            f"the training data file {os.path.abspath(filename)}. "
             f"Rasa Open Source will read the file as a "
             f"version '{LATEST_TRAINING_DATA_FORMAT_VERSION}' file. "
-            f"See {DOCS_BASE_URL}."
+            f"See {DOCS_URL_TRAINING_DATA}."
```
Comment (wochinge) [inline rasa/shared/utils/validation.py:221]: 🚀

**Review by wochinge — APPROVED** [review]
Thanks for providing such a detailed pull request description despite the lack of time! 
I approve to not block you, but would be great to quickly discuss the boolean checker functions (`is_story_file` etc)

**Comment by wochinge** [comment]
cc @federicotdn I think this change could have some Rasa X implications?

File: rasa/cli/test.py (hunk 3 above)
Comment (tmbo) [reply rasa/cli/test.py:142]: yes good, point :+1:

File: rasa/shared/core/domain.py (hunk 6 above)
Comment (tmbo) [reply rasa/shared/core/domain.py:1200]: yes I agree, it is not ideal. but just returning `False` will mean invalid files just get ignored, which isn't really helpful either. ideally we'd split test if a file is valid yaml before calling `is_domain` / `is_story` / ...

Comment (wochinge) [reply rasa/shared/core/domain.py:1200]: or we make all the boolean checks independent of yaml reading (test against the text directly)?

Comment (wochinge) [reply rasa/shared/core/domain.py:1200]: feel free to move forward but please create an issue to find a more robust solution after the release.

File: rasa/nlu/classifiers/sklearn_intent_classifier.py (hunk 4 above)
Comment (tmbo) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py:103]: :+1:

File: rasa/shared/exceptions.py (hunk 13 above)
Comment (tmbo) [reply rasa/shared/exceptions.py:55]: should we rather link to https://yamlchecker.com/ ?

Comment (wochinge) [reply rasa/shared/exceptions.py:55]: yes please! I also like that it gives instant feedback and I don't have to press `go`. Also it checks for duplicate keys and doesn't transform the yaml in a weird way.

File: rasa/nlu/classifiers/sklearn_intent_classifier.py (hunk 4 above)
Comment (wochinge) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:103]: this should use `DOCS_URL_TRAINING_DATA_NLU` then (same for the others)

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py [hunk 20]
```diff
@@ -139,6 +135,9 @@ def read_from_parsed_yaml(
     def is_stories_file(cls, file_path: Text) -> bool:
         """Check if file contains Core training data or rule data in YAML format.
 
+        Will throw exceptions if the file seems to be a yaml file (extension) but
```
Comment (wochinge) [inline rasa/shared/core/training_data/story_reader/yaml_story_reader.py:138]: The google docstring define a `Raises:` syntax https://google.github.io/styleguide/pyguide.html#244-decision

Comment (tmbo) [reply rasa/shared/core/training_data/story_reader/yaml_story_reader.py:138]: ah nice, wasn't aware of that :+1:

**Pull request merged**